### ISSUE TITLE
undef &$coderef consistency

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10012]
Migrated from [rt.perl.org#71154](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=71154) (status was 'resolved')

Searchable as RT71154$

**Comment by p5pRT** [comment]
### From @nwc10

#### Created by @nwc10



$ \./perl \-e '$a = sub \{\}; warn "Before&#8203;: $a"; undef &$a; warn "After $a"; $a\->\(\)'
Before&#8203;: CODE\(0x100815848\) at \-e line 1\.
After CODE\(0x100815848\) at \-e line 1\.
Not a CODE reference at \-e line 1\.

only it sure does look like a CODE reference\. Even to Devel&#8203;::Peek&#8203;:

$ \./perl \-Ilib \-MDevel&#8203;::Peek \-e '$a = sub \{\}; undef &$a; Dump $a'
SV = IV\(0x100815858\) at 0x100815860
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(ROK\)
&nbsp;  RV = 0x100815878
&nbsp;  SV = PVCV\(0x100814820\) at 0x100815878
&nbsp;    REFCNT = 2
&nbsp;    FLAGS = \(PADMY\,WEAKOUTSIDE\)
&nbsp;    COMP\_STASH = 0x100803d38    "main"
&nbsp;    ROOT = 0x0
&nbsp;    GVGV&#8203;::GV = 0x100803ff0      "main" :&#8203;: "\_\_ANON\_\_"
&nbsp;    FILE = "\(null\)"
&nbsp;    DEPTH = 0
&nbsp;    FLAGS = 0x10
&nbsp;    OUTSIDE\_SEQ = 96
&nbsp;    PADLIST = 0x0
&nbsp;    OUTSIDE = 0x1008040e0 \(MAIN\)

I'm not sure what the right answer here is\.

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.11.2:

Configured by nick at Tue Dec  8 18:23:09 GMT 2009.

Summary of my perl5 (revision 5 version 11 subversion 2) configuration:
  Commit id: bb61cbbfe4f34b8082fca34dd9dd82cae6aaa146
  Platform:
    osname=darwin, osvers=10.0.0, archname=darwin-thread-multi-2level
    uname='darwin mouse-mill.local 10.0.0 darwin kernel version 10.0.0: fri jul 31 22:47:34 pdt 2009; root:xnu-1456.1.25~1release_i386 i386 '
    config_args='-Dusedevel=y -Dcc=ccache gcc-mp-4.4 -Dld=gcc-mp-4.4 -Ubincompat5005 -Uinstallusrbinperl -Dcf_email=nick@ccl4.org -Dperladmin=nick@ccl4.org -Dinc_version_list=  -Dinc_version_list_init=0 -Doptimize=-Os -Dusethreads -Accflags=-DNO_MATHOMS -Uusemymalloc -Duseperlio -Dprefix=~/Sandpit/snap5.9.x-v5.11.2-138-gbb61cbb -Dinstallman1dir=none -Dinstallman3dir=none -Uuserelocatableinc -Ud_dosuid -Uuseshrplib -Umad -de -Accccflags=-DNO_PERL_PRESERVE_IVUV'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='ccache gcc-mp-4.4', ccflags ='-fno-common -DPERL_DARWIN -no-cpp-precomp -DNO_MATHOMS -fno-strict-aliasing -pipe -fstack-protector -I/opt/local/include',
    optimize='-Os',
    cppflags='-no-cpp-precomp -fno-common -DPERL_DARWIN -no-cpp-precomp -DNO_MATHOMS -fno-strict-aliasing -pipe -fstack-protector -I/opt/local/include'
    ccversion='', gccversion='4.4.1', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags =' -fstack-protector -L/opt/local/lib'
    libpth=/opt/local/lib /usr/lib
    libs=-lgdbm -ldbm -ldl -lm -lutil -lc
    perllibs=-ldl -lm -lutil -lc
    libc=/usr/lib/libc.dylib, so=dylib, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup -L/opt/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.11.2:
    lib
    /Users/nick/Sandpit/snap5.9.x-v5.11.2-138-gbb61cbb/lib/perl5/site_perl/5.11.2/darwin-thread-multi-2level
    /Users/nick/Sandpit/snap5.9.x-v5.11.2-138-gbb61cbb/lib/perl5/site_perl/5.11.2
    /Users/nick/Sandpit/snap5.9.x-v5.11.2-138-gbb61cbb/lib/perl5/5.11.2/darwin-thread-multi-2level
    /Users/nick/Sandpit/snap5.9.x-v5.11.2-138-gbb61cbb/lib/perl5/5.11.2
    .


Environment for perl 5.11.2:
    DYLD_LIBRARY_PATH (unset)
    HOME=/Users/nick
    LANG=en_GB.ISO8859-1
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/opt/local/bin:/opt/local/sbin:/Users/nick/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/usr/X11/bin:/usr/local/sbin:/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

On Wed\, Dec 09\, 2009 at 02&#8203;:42&#8203;:07PM \-0800\, Nicholas Clark wrote&#8203;:
> $ \./perl \-e '$a = sub \{\}; warn "Before&#8203;: $a"; undef &$a; warn "After $a"; $a\->\(\)'
> Before&#8203;: CODE\(0x100815848\) at \-e line 1\.
> After CODE\(0x100815848\) at \-e line 1\.
> Not a CODE reference at \-e line 1\.

A simple patch&#8203;:

\-\-\- a/pp\_hot\.c
\+\+\+ b/pp\_hot\.c
@&#8203;@&#8203; \-2738\,7 \+2738\,7 @&#8203;@&#8203; PP\(pp\_entersub\)
&nbsp;            DIE\(aTHX\_ "Undefined subroutine called"\);
 
&nbsp;        /\* autoloaded stub? \*/
\-       if \(cv \!= GvCV\(gv\)\) \{
\+       if \(cv \!= GvCV\(gv\) && GvCV\(gv\)\) \{
&nbsp;            cv = GvCV\(gv\);
&nbsp;        \}
&nbsp;        /\* should call AUTOLOAD now? \*/

\(which I don't recommend\) will change the error to the more
consistent&#8203;:

Before&#8203;: CODE\(0x1d0d8b8\) at \-e line 1\.
After CODE\(0x1d0d8b8\) at \-e line 1\.
Undefined subroutine &main&#8203;::\_\_ANON\_\_ called at \-e line 1\.

and still passes all tests\.

Compare to&#8203;:

tony@&#8203;mars&#8203;:\.\.\./git/perl$ perl \-le 'sub foo \{\} $a = \\&foo; warn "Before $a"; undef &$a; warn "After $a"; $a\->\(\)'
Before CODE\(0x1fc6f00\) at \-e line 1\.
After CODE\(0x1fc6f00\) at \-e line 1\.
Undefined subroutine &main&#8203;::foo called at \-e line 1\.

\(I suspect with the patch it tries to AUTOLOAD \_\_ANON\_\_\.\)

What would the desired behaviour be?

Tony

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Dec 10 05&#8203;:31&#8203;:21 2009\, tonyc wrote&#8203;:
> On Wed\, Dec 09\, 2009 at 02&#8203;:42&#8203;:07PM \-0800\, Nicholas Clark wrote&#8203;:
> > $ \./perl \-e '$a = sub \{\}; warn "Before&#8203;: $a"; undef &$a; warn "After
> $a"; $a\->\(\)'
> > Before&#8203;: CODE\(0x100815848\) at \-e line 1\.
> > After CODE\(0x100815848\) at \-e line 1\.
> > Not a CODE reference at \-e line 1\.
> 
> A simple patch&#8203;:
> 
\.\.\.

> \(which I don't recommend\) will change the error to the more
> consistent&#8203;:
> 
> Before&#8203;: CODE\(0x1d0d8b8\) at \-e line 1\.
> After CODE\(0x1d0d8b8\) at \-e line 1\.
> Undefined subroutine &main&#8203;::\_\_ANON\_\_ called at \-e line 1\.

Just a simple ‘Undefined subroutine called’ would be appropriate\, would
it not?

**Comment by p5pRT** [comment]
### From @cpansprout

Now fixed with commit 2c3743704\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Dec 09 14&#8203;:42&#8203;:07 2009\, nicholas wrote&#8203;:
> $ \./perl \-e '$a = sub \{\}; warn "Before&#8203;: $a"; undef &$a; warn "After
>    $a"; $a\->\(\)'
> Before&#8203;: CODE\(0x100815848\) at \-e line 1\.
> After CODE\(0x100815848\) at \-e line 1\.
> Not a CODE reference at \-e line 1\.
> 
> only it sure does look like a CODE reference\.

I fixed that to produce a sensible error message in commit 2c3743704\,
which is in 5\.16\.

What I did not realise was that the problem actually went further \(and
my commit fixed this\, too\)&#8203;:

sub \_\_ANON\_\_ \{ warn 42 \}
$x = sub \{\};
undef &$x;
$x\->\(\);

With 5\.14\.0&#8203;:
42 at \- line 1\.

With 5\.16\.0&#8203;:
Undefined subroutine called at \- line 4\.

So I have improved the test with commit 12f98b43fb8\, which makes sure
\_\_ANON\_\_\(\) is not called\.

But the problem goes even deeper than that\.  This ‘Not a CODE
reference’ error was occurring whenever an already\-autoloaded subroutine
was looked for when calling a stub\, if the GV had no CV in it\.  Compare&#8203;:

$ perl5\.16\.0 \-e '
&nbsp;  my $foosub = \\&foo;
&nbsp;  &$foosub;
'
Undefined subroutine &main&#8203;::foo called at \-e line 3\.

$ perl5\.16\.0 \-e '
&nbsp;  my $foosub = \\&foo;
&nbsp;  undef \*foo;
&nbsp;  &$foosub;
'
Not a CODE reference at \-e line 4\.

So I have just fixed that now with commit c220e1a11e\.

\-\- 

Father Chrysostomos
